Update Readme file (new template) (#6)
DevExpress-Examples/wpf-scheduler-apply-end-user-restrictions: Readme.md

[![](https://img.shields.io/badge/💬_Leave_Feedback-feecdd?style=flat-square)](#does-this-example-address-your-development-requirementsobjectives)
<!-- default badges end -->

# WPF Scheduler - Apply User Restrictions

This example applies the following user restrictions:

* Prevents users from creating new appointments.

> * Prevents users from moving appointments into the lunch break interval (12:00–13:00).

Reviewer: ```suggestion
* Prevents users from moving appointments to the lunch break interval (12:00–13:00).
```

![Apply User Restrictions - WPF Scheduler, DevExpress](./Images/scheduler.jpg)

## Implementation Details

### Define Restricted Interval

In this example, the restricted interval covers the lunch break time (12:00–13:00):

```csharp
TimeInterval lunchTime = new TimeInterval(DateTime.Today.AddHours(12), TimeSpan.FromHours(1));
```

The `IsIntervalAllowed` method returns `false` if any part of the selected interval overlaps with the restricted time range:

```csharp
private bool IsIntervalAllowed(TimeInterval interval) {
    DateTime dayStart = interval.Start.Date;
    while (dayStart < interval.End) {
        if (interval.IntersectsWithExcludingBounds(lunchTime))
            return false;
        dayStart = dayStart.AddDays(1);
    }
    return true;
}
```

### Restrict Appointment Creation

The [`CustomAllowAppointmentCreate`](https://docs.devexpress.com/WPF/DevExpress.Xpf.Scheduling.SchedulerControl.CustomAllowAppointmentCreate) event allows you to define rules that restrict appointment creation. In this example, the event handler:

* Blocks appointment creation if **Disable Creating Appointments** is turned on.

> * Prevents users from creating appointments during the launch time if **Disable Creating Appointments** is turned off.

Reviewer: ```suggestion
* Prevents users from creating appointments during lunch time if **Disable Creating Appointments** is turned off.
```

```csharp
private void customAllowAppointmentCreateHandler(object sender, AppointmentItemOperationEventArgs e) {
    // Cancel the operation if the "Disable Creating Appointments" bar item is checked
    if((bool)barItemDisableCreatingAppointments.IsChecked) {
        e.Allow = false;
        return;
    }
    DateTimeRange selectedIntervalRange = schedulerControl1.SelectedInterval;
    TimeInterval selectedInterval = new TimeInterval(selectedIntervalRange.Start, selectedIntervalRange.End);
    e.Allow = IsIntervalAllowed(selectedInterval);
}
```

The `barItemDisableCreatingAppointments.IsChecked` flag is bound to the **Disable Creating Appointments** Ribbon item and acts as a switch.

### Resolve Appointment Conflicts

> The [CustomAllowAppointmentConflicts](https://docs.devexpress.com/WPF/DevExpress.Xpf.Scheduling.SchedulerControl.CustomAllowAppointmentConflicts) event allows you to control how the scheduler handles appointment conflicts. In this example, a conflict occurs when a user moves an appointment into a time interval that already contains another appointment or into the lunch break. In the Ribbon UI, you can enable the **Disable Appointment Conflicts** item to ignore conflicts.

Reviewer: ```suggestion
The [CustomAllowAppointmentConflicts](https://docs.devexpress.com/WPF/DevExpress.Xpf.Scheduling.SchedulerControl.CustomAllowAppointmentConflicts) event allows you to control how the scheduler handles appointment conflicts. In this example, a conflict occurs when a user moves an appointment to a time interval that already contains another appointment or to the lunch break time. In the Ribbon UI, you can enable the **Disable Appointment Conflicts** item to ignore conflicts.
```
